### PR TITLE
Nan in linalg eig

### LINKED ISSUE #61251: torch.linalg.eig() doesn't handle NaNs
## 🐛 Bug
This is a followup to https://github.com/pytorch/pytorch/issues/37499. As it was decided to handle NaNs in `x.eig()` for a tensor `x` there, I suppose they should also be handled for `torch.linalg.eig(x)`.

### To reproduce
```
import torch
import numpy as np
A = np.nan * torch.ones((3,3))
torch.linalg.eig(A)
```
This currently gives:
```
Intel MKL ERROR: Parameter 3 was incorrect on entry to SGEBAL.
Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)
```
### Environment
 - PyTorch Version: 1.9.0+cu102
 - OS (e.g., Linux): Ubuntu 20.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Python version: 3.7.7
 - CUDA/cuDNN version: 10.2

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @walterddr @IvanYashchuk @xwang233 @Lezcano

### PR DESCRIPTION
**Pull request by v0dro**
Fixes https://github.com/pytorch/pytorch/issues/61251. As per the comment here (https://github.com/pytorch/pytorch/issues/61251#issuecomment-954676082), a consensus has been reached to raise an error if there is a NaN value in the input when calling `eig()`. This PR implements that feature.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

### CODE REVIEW

**Comment by v0dro** [comment]
The Vulkan test failure seems to be unrelated.

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp [hunk 1]
```diff
@@ -261,7 +262,7 @@ void apply_linalg_eig(Tensor& values, Tensor& vectors, Tensor& input, Tensor& in
     scalar_t* rvectors_working_ptr = compute_eigenvectors ? &rvectors_data[i * input_matrix_stride] : nullptr;
     int* info_working_ptr = &infos_data[i];
     lapackEig<scalar_t, value_t>(jobvl, jobvr, n, input_working_ptr, lda, values_working_ptr,
-      lvectors_data, ldvl, rvectors_working_ptr, ldvr, work_data, lwork, rwork_data, info_working_ptr);
+                                 lvectors_data, ldvl, rvectors_working_ptr, ldvr, work_data, lwork, rwork_data, info_working_ptr);
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:265]: Changes in this file are not related, they can be reverted with `git checkout master -- aten/src/ATen/native/BatchLinearAlgebraKernel.cpp`

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -2544,6 +2544,7 @@ void apply_linalg_eig(Tensor& values, Tensor& vectors, Tensor& input, Tensor& in
 TORCH_CHECK(false, "Calling torch.linalg.eig on a CUDA tensor requires compiling PyTorch with MAGMA. "
                    "Either transfer the tensor to the CPU before calling torch.linalg.eig or recompile with MAGMA.");
 #else
+ TORCH_CHECK(input.isfinite().all().item<bool>(), "apply_linalg_eig: input tensor should not contain infs or NaNs.");
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:2547]: This check is not needed here because it's already included higher in the call hierarchy.

File: test/test_linalg.py [hunk 3]
```diff
@@ -2295,6 +2295,21 @@ def test_eig_errors_and_warnings(self, device, dtype):
             with self.assertRaisesRegex(RuntimeError, "tensors to be on the same device"):
                 torch.linalg.eig(a, out=(out_w, out_v))
 
+    @skipCPUIfNoLapack
+    @dtypes(*floating_and_complex_types())
+    def test_eig_with_nan(self, device, dtype):
```
Comment (IvanYashchuk) [inline test/test_linalg.py:2300]: This test is fine as it is, but I think it can be rewritten to contain fewer lines with a for loop, something like:
```
for batch_dim in [(), (10,)]:
    a = make_tensor((*batch_dim, 5, 5), device=device, dtype=dtype)
    a[..., -1, -1] = float('nan')
    with self.assertRaisesRegex(...):
        torch.linalg.eig(a)
```

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 4]
```diff
@@ -2799,6 +2800,8 @@ std::tuple<Tensor&, Tensor&> linalg_eig_out(const Tensor& input, Tensor& values,
 }
 
 std::tuple<Tensor, Tensor> linalg_eig(const Tensor& input) {
+  TORCH_CHECK(input.isfinite().all().item<bool>(), "linalg_eig: input tensor should not contain infs or NaNs.");
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:2803]: `linalg_eig` functions calls to `linalg_eig_out`. Could you please move this check to the `linalg_eig_out` function?

Comment (IvanYashchuk) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:2803]: It's better to use user-facing Python function names. See other errors in `linalg_eig_out`.
```suggestion
  TORCH_CHECK(input.isfinite().all().item<bool>(), "torch.linalg.eig: input tensor should not contain infs or NaNs.");
```

File: test/test_linalg.py (hunk 3 above)
Comment (IvanYashchuk) [reply test/test_linalg.py:2300]: There's no test for inputs containing infinity values.

**Review by IvanYashchuk — CHANGES_REQUESTED** [review]
Thank you, Sameer, for submitting the fix that prevents crashes. In the error message, you added it's told that infinity is not allowed, but there are no tests for it.

**Comment by v0dro** [comment]
Thanks for the review @IvanYashchuk ! I've updated the tests and error as per your feedback. Indeed it is more concise now. Do let me if this new iteration is alright.

File: test/test_linalg.py [hunk 5]
```diff
@@ -2295,6 +2295,18 @@ def test_eig_errors_and_warnings(self, device, dtype):
             with self.assertRaisesRegex(RuntimeError, "tensors to be on the same device"):
                 torch.linalg.eig(a, out=(out_w, out_v))
 
+    @skipCPUIfNoLapack
+    @skipCUDAIfNoMagma
+    @dtypes(*floating_and_complex_types())
+    def test_eig_with_nan(self, device, dtype):
+        for val in [np.inf, np.nan]:
```
Comment (nikitaved) [inline test/test_linalg.py:2302]: Is it allright to use numpy here? Afaik numpy is no longer in the dependency, right? Probably it is better to replace it with `float('inf')` and `float('nan')`. Also, I am not sure this approach works well with float32, as default floating type for numpy/python is double.

Comment (Lezcano) [reply test/test_linalg.py:2302]: That being said, I think the `float` solution is better, as `np` is not necessary here. Just a thought tho, not even a nit.

Comment (nikitaved) [reply test/test_linalg.py:2302]: Yet I am not sure this test is correct for 32-bit floats.

**Review by IvanYashchuk — APPROVED** [review]

File: test/test_linalg.py (hunk 5 above)
Comment (mruberry) [reply test/test_linalg.py:2302]: NumPy is no longer a dependency of PyTorch itself but we do require it to run the tests, so yes you're OK to use NumPy unconditionally in the test suite

**Review by mruberry — APPROVED** [review]
Cool!

Performance impact not a concern as per linear algebra meeting.

Thanks @v0dro!

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 6]
```diff
@@ -2712,6 +2712,7 @@ std::tuple<Tensor&, Tensor&> linalg_eig_out_info(const Tensor& input, Tensor& va
 }
 
 std::tuple<Tensor&, Tensor&> linalg_eig_out(const Tensor& input, Tensor& values, Tensor& vectors) {
+  TORCH_CHECK(input.isfinite().all().item<bool>(), "torch.linalg.eig: input tensor should not contain infs or NaNs.");
```
[the remark this answers is not in the document]
Comment (mruberry) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:2715]: Update: not an issue per linear algebra meeting.

[the remark this answers is not in the document]
Comment (xwang233) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:2715]: Thanks for the work! I think the performance hit is tolerable and it's no longer a performance concern to me. 😄
